src/tests/JIT/Directed/Convert/out_of_range_fp_to_int_conversions.cs

```diff
@@ -0,0 +1,448 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Numerics;
+
+namespace FPBehaviorApp
+{
+    public enum FPtoIntegerConversionType
+    {
+        CONVERT_BACKWARD_COMPATIBLE,
+        CONVERT_SENTINEL,
+        CONVERT_SATURATING,
+        CONVERT_NATIVECOMPILERBEHAVIOR,
+        CONVERT_MANAGED_BACKWARD_COMPATIBLE_X86_X64,
+        CONVERT_MANAGED_BACKWARD_COMPATIBLE_ARM32,
+    }
+
+    public enum ConversionType
+    {
+        ToInt,
+        ToLong,
+        ToUInt,
+        ToULong
+    }
+
+    public static class Native
+    {
+        [DllImport("out_of_range_fp_to_int_conversionsnative")]
+        [SuppressGCTransition]
+        public static extern int ConvertDoubleToLong(double x, FPtoIntegerConversionType t);
+
+        [DllImport("out_of_range_fp_to_int_conversionsnative")]
+        [SuppressGCTransition]
+        public static extern uint ConvertDoubleToUnsignedLong(double x, FPtoIntegerConversionType t);
+
+        [DllImport("out_of_range_fp_to_int_conversionsnative")]
+        [SuppressGCTransition]
+        public static extern long ConvertDoubleToLongLong(double x, FPtoIntegerConversionType t);
+
+        [DllImport("out_of_range_fp_to_int_conversionsnative")]
+        [SuppressGCTransition]
+        public static extern ulong ConvertDoubleToUnsignedLongLong(double x, FPtoIntegerConversionType t);
+    }
+
+    public static class Managed
+    {
+        public static void DumpBits(double d)
+        {
+            ulong uintVal = (ulong)BitConverter.DoubleToInt64Bits(d);
+            bool signBit = (int)((uintVal >> 63) & 0x1) != 0;
+            ulong mantissa = uintVal & 0x000F_FFFF_FFFF_FFFF;
+            int exponent = (int)((uintVal >> 52) & 0x7FF);
+            Console.WriteLine($"{signBit} {exponent:x} {mantissa:x}");
+        }
+
+        // Equivalent to Math.Truncate, but written in C# to avoid Debug/Check penalties running on CoreCLR test builds
```

Comment:
What are these penalties?

Reply:
There are a long series of assertions that run when an FCALL is made using a chk or debug build of the runtime. In general they don't matter much for performance, but when I was running experiments hitting 1,000,000,000 different values to see if I had any holes in my work, the cost of the assertions added up fast.